Language/Statements/Local_Variable_Declaration/declared_type_A03_t01.dart

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion The declared type of a local variable with a declaration of one
+/// of the forms late? var v = e; late? final v = e; const v = e; is determined
+/// as follows:
+/// ...
+/// • If the static type of e is of the form X & T where X is a type variable,
+/// the declared type of v is X. In this case v is immediately promoted to X & T
+///
+/// @description Checks that static type of a variable declared by the
+/// statements `late? var v = e; late? final v = e;` is `X` if the static type
+/// of `e` is `X & T` where `X` is a type variable. Also test that `v` is
+/// promoted to `X & T`
+/// @author sgrekhov22@gmail.com
+
+import '../../../Utils/static_type_helper.dart';
+
+test1<T>(T t) {
+  if (t is int) {
+    var v = t;
+    v.isEven;
+    v.expectStaticType<Exactly<T>>();
+  }
+  var v = t;
+  v.expectStaticType<Exactly<T>>();
+}
+
+test2<T>(T t) {
+  if (t is int) {
+    late var v = t;
+    v.isEven;
+    v.expectStaticType<Exactly<T>>();
+  }
+  late var v = t;
+  v.expectStaticType<Exactly<T>>();
+}
+
+test3<T>(T t) {
+  if (t is int) {
+    final v = t;
```

Comment:
This is an interesting case. 

Double checking with the NNBD feature specification (the updates in the language specification shouldn't disagree with the NNBD feature specification except where something has been changed intentionally, which is basically small bug fixes --- and I don't think we had any of those here):

In the NNBD feature specification, it is not stated explicitly that the promotion to `X & T` is not applied to a `final` local variable when the initializing expression has type `X & T`, but it could be considered to be implied by the reference to `final` a few lines earlier.

However, the implementations _do_ actually promote `final v = e;` in this way, so it would be a breaking change to say that there is no promotion in this case.

It's a weird corner case ("why do we promote `final x = e;` from `X` to `X & int` when `e` has static type `X & int`, but we do _not_ promote `final int? x = e;` to `int` when `e` has static type `int`?"), but I don't see a clear discrepancy, and the promotion is already implemented. (Also, I don't see an overwhelming need to broaden the special exception for `final` variables.)

So it's fine, and we _should_ expect `v` to be promoted to `T & int` here! ;-)

So we still have the same considerations about testing `T & int`.